readium/shared/src/main/java/org/readium/r2/shared/publication/asset/RemoteAsset.kt

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2023 Readium Foundation. All rights reserved.
+ * Use of this source code is governed by the BSD-style license
+ * available in the top-level LICENSE file of the project.
+ */
+
+package org.readium.r2.shared.publication.asset
+
+import java.net.URL
+import org.readium.r2.shared.extensions.extension
+import org.readium.r2.shared.fetcher.ArchiveFetcher
+import org.readium.r2.shared.fetcher.Fetcher
+import org.readium.r2.shared.fetcher.HttpFetcher
+import org.readium.r2.shared.publication.Link
+import org.readium.r2.shared.publication.Publication
+import org.readium.r2.shared.util.Try
+import org.readium.r2.shared.util.mediatype.MediaType
+
+/**
+ * Represents a publication accessible remotely.
+ */
+class RemoteAsset(
+    val url: URL,
+    private val knownMediaType: MediaType?,
+    private val mediaTypeHint: String?
+) : PublicationAsset {
+
+    /**
+     * Creates a [RemoteAsset] from a [URL] and an optional media type, when known.
+     */
+    constructor(url: URL, mediaType: MediaType? = null) :
+        this(url, knownMediaType = mediaType, mediaTypeHint = null)
+
+    /**
+     * Creates a [RemoteAsset] from a [URL] and an optional media type hint.
+     *
+     * Providing a media type hint will improve performances when sniffing the media type.
+     */
+    constructor(url: URL, mediaTypeHint: String?) :
+        this(url, knownMediaType = null, mediaTypeHint = mediaTypeHint)
+
+    override suspend fun mediaType(): MediaType {
+        if (!::_mediaType.isInitialized) {
+            val bytes = { url.openConnection().getInputStream().use { it.readBytes() } }
```

Comment:
There's a [`URL.readBytes()`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/java.net.-u-r-l/read-bytes.html) extension in the Kotlin stdlib which basically does the same thing.